rcl/test/rcl/test_graph.cpp

```diff
@@ -1295,6 +1295,32 @@ TEST_F(CLASSNAME(TestGraphFixture, RMW_IMPLEMENTATION), test_graph_guard_conditi
   ASSERT_GE(graph_changes_count, 4ul);
 }
 
+/* Test the graph when guard condition is triggered.
+ */
+TEST_F(CLASSNAME(TestGraphFixture, RMW_IMPLEMENTATION), test_graph_guard_condition_triggered) {
+  rcl_ret_t ret;
+  int64_t wait_time_out = -1;
+  const rcl_guard_condition_t * graph_guard_condition =
+    rcl_node_get_graph_guard_condition(this->node_ptr);
+  ASSERT_NE(nullptr, graph_guard_condition) << rcl_get_error_string().str;
+  ret = rcl_wait_set_clear(this->wait_set_ptr);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  ret = rcl_wait_set_add_guard_condition(this->wait_set_ptr, graph_guard_condition, NULL);
+  ASSERT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+
+  std::thread trigger_thread(
+    [ = ]() {
+      std::this_thread::sleep_for(std::chrono::milliseconds(200));
+      rcl_ret_t ret =
+      rcl_trigger_guard_condition(const_cast<rcl_guard_condition_t *>(graph_guard_condition));
+      EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+    });
+
+  ret = rcl_wait(this->wait_set_ptr, wait_time_out);
+  EXPECT_EQ(RCL_RET_OK, ret) << rcl_get_error_string().str;
+  trigger_thread.join();
```

Comment:
What is the higher level purpose of this test? From my perspective it's essentially testing that triggering a guard condition will wake up a wait set, but is the goal to do something specific to the graph guard condition?

Reading https://github.com/ros2/rcl/issues/574, I don't think this test really addresses it, because it doesn't check that the graph guard condition is triggered by certain events, nor does it assert anything about behavior that utilizes the graph guard condition (e.g. the GraphListener in rclcpp).

@ivanpauno perhaps you could explain what kind of test you were expecting from https://github.com/ros2/rcl/issues/574?

Reply:
> @ivanpauno perhaps you could explain what kind of test you were expecting from #574?

Sorry, my description of the issue in #574 is a bit poor.
AFAIR, my motivation for opening that issue was that some rmw implementations weren't triggering the graph guard condition in all graph changes.
My idea was to add more tests that check that after each function that can have a "graph side effect", the graph guard conditions is triggered within a timeout.
e.g.:
```
-> create a node
  -> check that graph guard condition is triggered within a timeout
-> create a publisher
  -> check that graph guard condition is triggered within a timeout
... etc for everything that can produce a graph change
```

I think that the only existing test is [this one]( https://github.com/ros2/rcl/blob/2daa408ea46bb2a4305a6221d73ea400e4c02d62/rcl/test/rcl/test_graph.cpp#L1231).
That test is a bit fragile, but my basic idea is to make that more extensive to cover all graph changes.
There are some cases that are hard to test, i.e. making sure that a graph guard condition is triggered when a publisher matching an existing subscription is detected (in that case the graph guard condition is triggered twice, once because the publisher was detected and the other because the subscription found a matching publisher).

> What is the higher level purpose of this test? 

I also don't see value in the test added here.

Reply:
I have edited #574 to make its purpose more clear.